certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,240 @@
+    front()             returns (bytes32) envfree
+    back()              returns (bytes32) envfree
+    at_(uint256)        returns (bytes32) envfree // at is a reserved word
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Helpers                                                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+function absolute_diff(mathint a, mathint b) returns mathint {
+    return a > b ? a - b : b - a; 
+}
+
+function min_int128() returns mathint {
+    return max_uint256 / 2 + 1;
+}
+
+function max_int128() returns mathint {
+    return max_uint256 / 2;
+}
+
+// Could be broken in theory, but not in practice
+function beginLimited(int128 begin) returns bool {
+    return min_int128() < to_mathint(begin); // not inclusive so it can pushFront
+}
+
+// Could be broken in theory, but not in practice
+function endLimited(int128 end) returns bool {
+    return max_int128() > to_mathint(end); // not inclusive so it can pushBack
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: end is larger or equal than begin if values are safely assumed not to overflow.                          │
+| NOTE: This assumption make the proofs unsound, but they can't realistically overflow                                |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant boundariesConsistency()
+    end() >= begin()
+    {
+        preserved {
+            require beginLimited(begin());
+            require endLimited(end());
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: length is end minus begin                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant lengthConsistency()
+    length() == absolute_diff(to_mathint(end()), to_mathint(begin()))
+    {
+        preserved {
+            requireInvariant boundariesConsistency;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: out of bounds is always 0                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant outOfBoundsDefault(int128 key)
+    (
+        key < begin() &&  // Is before beginning
+        key >= end() // Is after ending
+    ) => at_(to_uint256(key)) == 0
```

Why not directly work with a uint256 key ?

```suggestion
invariant outOfBoundsDefault(uint256 key)
    (
        key < begin() ||  // Is before beginning
        key >= end() // Is after ending
    ) => at_(key) == 0
```

---

Also, I think the `&&` should be a `||`  !!!

---

Last but not least, you are confusing the `uint256` position for `at` with the `int256` position in the storage mapping. Lot at the code for `at`, we apply an offset. 

This rule should be something like 
```key >= length <=> at_(key) reverts```
which you could write as
```key < length <=> at_(key) >= 0```
(`at_(key) >= 0` being a way to say `at_(key)` returns something)

---

Both are correct, I left multiple mistakes trying to make everything fit. Should be clearer now.